toolchain/lower/function_context.cpp

```diff
@@ -153,9 +153,16 @@ auto FunctionContext::GetDebugLoc(SemIR::InstId inst_id) -> llvm::DebugLoc {
     return llvm::DebugLoc();
   }
   auto loc = file_context_->GetLocForDI(inst_id);
-  CARBON_CHECK(loc.filename == di_subprogram_->getFile()->getFilename(),
-               "Instructions located in a different file from their "
-               "enclosing function aren't handled yet");
+  if (loc.filename != di_subprogram_->getFile()->getFilename()) {
+    // Location is from a different file. We can't represent that directly
+    // within the scope of this function's subprogram, and we don't want to
+    // generate a new subprogram, so just discard the location information. This
+    // happens for thunks when emitting the portion of the thunk that is
+    // duplicated from the original signature.
+    //
+    // TODO: Handle this case better.
```

Comment:
Yeah, not clear what the representation is that we'd want - the way Clang/LLVM IR handles C++ macros, for instance, is to create a `DILexicalBlockFile` to represent a change in file (the resulting DWARF describes all the instructions as being in the same function, but some instructions are attributed to one file, and some to another, in the DWARF line table (.debug_line)) - alternatively there's inlined functions, but that doesn't seem quite like how we'd wan tto describe what's going on here. Clang's logic for this is here: https://github.com/llvm/llvm-project/blob/a2b28a68125c5e179df01ed62ca8549bcb9c73b1/clang/lib/CodeGen/CGDebugInfo.cpp#L228-L240